src/import-shadow-hash.c

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "storage.h"
+#include "module.h"
+#include "entry.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <shadow.h>
+
+/**
+ * CreateEntryFromShadowHash:
+ * @username: Username containing the shadow hash.
+ * @sdhash: User's hashed shadow password.
+ */
+gboolean CreateEntryFromShadowHash(const char *username, char *sdhash) {
+  // Create a new Cache Entry.
+  CacheEntry *entry = CacheEntryNew();
+  GBytes* gbytes = g_bytes_new(sdhash, strlen(sdhash));
+  // Set the newly created entry hash using the /etc/shadow hash.
+  CacheEntryHashSet(entry, CACHE_ENTRY_ALGORITHM_CRYPT, gbytes, NULL);
+  g_bytes_unref(gbytes);
+  GError *error = NULL;
+  CacheStorage *storage = CacheStorageNew(DEFAULT_STORAGE_PATH);
```

Comment:
In the unit test call CacheTestInitFiles() from test.h, then use "/mock" for the path to CacheStorageNew(path) just like storage_test.c does. That writes files to a hashtable in memory instead of to disk.